### PR TITLE
DAOS-14441 test: Perform exclude or drain during/after extend rebuild process.

### PR DESCRIPTION
**Pull request by rpadma2**
Test-tag: test_osa_offline_extend_exclude_during_rebuild test_osa_offline_extend_drain_after_rebuild test_osa_online_extend_exclude_during_rebuild test_osa_online_extend_drain_after_rebuild

Summary:

- Add new test cases to existing OSA tests (exclude or drain during/after the extend operation respectively).
- Add the tests to offline and online OSA tests.
- Drain feature is not supported when extend rebuild happens. It can work only after extend rebuild is completed. It is not the case for exclude [ie: we can perform an exclude when extend rebuild happens].

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/osa/offline_extend.yaml [hunk 1]
```diff
@@ -93,3 +93,6 @@ snapshot:
   test_with_snapshot: true
 test_ranks:
   rank_list: ["4", "5"]
+perform_osa_tasks:
+  exclude: "exclude"
+  drain: "drain"
```
Comment (daltonbohning) [inline src/tests/ftest/osa/offline_extend.yaml:98]: What's the point of these if the value is the same as the key?

Comment (rpadma2) [reply src/tests/ftest/osa/offline_extend.yaml:98]: yes, will remove this option via YAML.

File: src/tests/ftest/osa/offline_extend.py [hunk 2]
```diff
@@ -27,9 +28,11 @@ def setUp(self):
             "test_servers", "server_partition", "server_reservation", "/run/extra_servers/*")
         self.rank = self.params.get("rank_list", "/run/test_ranks/*")
         self.test_oclass = None
+        self.test_exclude_or_drain = None
```
Comment (daltonbohning) [inline src/tests/ftest/osa/offline_extend.py:31]: This can be removed now :)

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/osa/offline_extend.py [hunk 3]
```diff
@@ -29,7 +30,8 @@ def setUp(self):
         self.test_oclass = None
         self.dmg_command.exit_status_exception = True
 
-    def run_offline_extend_test(self, num_pool, data=False, oclass=None):
+    def run_offline_extend_test(self, num_pool, data=False, oclass=None,
+                                exclude_or_drain=None):
```
Comment (shimizukko) [inline src/tests/ftest/osa/offline_extend.py:34]: You might want to add description for `exclude_or_drain` argument.

File: src/tests/ftest/osa/offline_extend.py [hunk 4]
```diff
@@ -95,6 +97,17 @@ def run_offline_extend_test(self, num_pool, data=False, oclass=None):
             if self.test_during_aggregation is True and (num_pool > 1):
                 self.delete_extra_container(self.pool)
             output = self.pool.extend(rank_val)
+            self.log.info(output)
+            if exclude_or_drain == "exclude":
+                self.pool.wait_for_rebuild_to_start()
+                sleep(4)
```
Comment (shimizukko) [inline src/tests/ftest/osa/offline_extend.py:103]: You might want to add comment describing why we need 4 sec sleep here.

**Review by shimizukko — DISMISSED** [review]

File: src/tests/ftest/osa/offline_extend.py (hunk 3 above)
Comment (rpadma2) [reply src/tests/ftest/osa/offline_extend.py:34]: done

File: src/tests/ftest/osa/offline_extend.py (hunk 4 above)
Comment (rpadma2) [reply src/tests/ftest/osa/offline_extend.py:106]: done.

**Review by shimizukko — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Comment by rpadma2** [comment]
@daos-stack/daos-gatekeeper  : This PR is ready for merge.

**Pull request merged**